### PR TITLE
Port adaptive_avg_pool3d to ATen

### LINKED ISSUE #18065: Port adaptive_avg_pool3d to ATen
adaptive_avg_pool3d is a neural network function that is currently implemented in our legacy THNN (CPU) / THCUNN (CUDA) libraries. It is generally better if these live in our new library ATen, since it is more feature complete and reduces cognitive overhead.

Info about how to port https://github.com/pytorch/pytorch/wiki/TH-to-ATen-porting-guide

Note that adaptive_avg_pool2d was ported in #14714

### PR DESCRIPTION
**Pull request by crcrpar**
Resolves #18065.

### CODE REVIEW

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+        adaptiveaveragepool_loop(
+            input_data, output_data,
+            totalZ,
+            isizeT, isizeH, isizeW,
+            osizeT, osizeH, osizeW,
+            istrideD, istrideT, istrideH, istrideW);
+      });
+}
+
+void adaptive_avg_pool3d_backward_out_cuda_template(
```
Comment (crcrpar) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: @skrah @xmnlab 
Since this CUDA backward test fails and I'm struggling to fix this, could you take a look?

best regards

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+  auto output = at::empty({0}, input.options());
+  adaptive_avg_pool3d_out_cuda_template(output, input, output_size);
+  return output;
+}
+
+Tensor& adaptive_avg_pool3d_backward_out_cuda(
+    Tensor& gradInput,
+    const Tensor& gradOutput_,
+    const Tensor& input) {
+  gradInput.resize_as_(input);
```
Comment (skrah) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: Resizing twice, also in the following function.

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+    isizeW = input.size(4);
+
+    osizeT = gradOutput.size(2);
+    osizeH = gradOutput.size(3);
+    osizeW = gradOutput.size(4);
+  }
+
+  // somehow nonatomic is passing all test for volumetric case. (copied from
+  // THCUNN source code.)
+  // bool atomic = (isizeW%osizeW != 0) || (isizeH%osizeH != 0) || // (isizeT%osizeT != 0)
```
Comment (skrah) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: Can you try re-enabling line 448?

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+  bool atomic = false;
+
+  if (input.ndimension() == 4) {
+    totalZ = sizeD * osizeT;
```
Comment (skrah) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: ... and restoring these?

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu (hunk 3 above)
Comment (skrah) [reply aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: I know you took it over from `THCUNN`, but it's enabled in `AdaptiveMaxPool3d` for instance.

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+          /* set output to local average */
+          *op = sum / kT / kH / kW;
+        }
+      }
+    }
+  }
+}
+
+void adaptive_avg_pool3d_out_cpu_template(
+    at::Tensor& output,
```
Comment (xmnlab) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: it is not a issue but just a comment, as it is also on the `at` namespace I think you can just use `Tensor` as you have use in other places in this file. but it is not something important .. just a comment

**Review by xmnlab — COMMENTED** [review]
as an overview it seems good. after you do the changes @skrah suggest, I will compare the old code with the new one to see if I find anything else.

**Comment by skrah** [comment]
Just to be clear, I only tried to address the test failures. I haven't reviewed yet.

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu (hunk 2 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: ah oh, thank you!

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu (hunk 3 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: I tried and it worked. So I chose line 448.

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp (hunk 5 above)
Comment (crcrpar) [reply aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: Thank you. I like this kind of consistency and made changes.

**Comment by crcrpar** [comment]
@skrah @xmnlab Thank you for your cooperation and now the problem seems resolved.

@ezyang could you take a look?

**Comment by skrah** [comment]
@crcrpar Good that the issue is resolved!  I think we were asked to do a full review, so I'll start that (perhaps on Monday).  So far I've only focused on that one function. :)

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    const Tensor& input) {
+  /* sizes */
+  int sizeD = input.size(-4);
+  int isizeT = input.size(-3);
+  int isizeH = input.size(-2);
+  int isizeW = input.size(-1);
+  int osizeT = gradOutput_.size(-3);
+  int osizeH = gradOutput_.size(-2);
+  int osizeW = gradOutput_.size(-1);
+
```
Comment (skrah) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: The old code had `int64_t` here.  `AdaptiveAveragePooling.cpp` seems to use `int`, too, but in general I'd hesitate to change integer sizes during porting.

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+  /* get contiguous gradOutput */
+  auto gradOutput = gradOutput_.contiguous();
```
Comment (skrah) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: Nitpick: The sizes should be the same, but I find it easier to use the sizes after conversion to contiguous, which is what the existing code did.

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+              isizeW,
+              osizeT,
+              osizeH,
+              osizeW);
+        });
+  } else {
+    int64_t b;
+#pragma omp parallel for private(b)
+    for (b = 0; b < input.size(0); b++) {
+      AT_DISPATCH_FLOATING_TYPES_AND_HALF(
```
Comment (skrah) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: The example PR for `AdaptiveAveragePooling` does it too, but actually dispatch should not happen inside the loop.

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+  auto output = at::empty({0}, input.options());
+  adaptive_avg_pool3d_out_cpu_template(output, input, output_size);
+  return output;
+}
+
+Tensor& adaptive_avg_pool3d_backward_out_cpu(
+    Tensor& gradInput,
+    const Tensor& gradOutput_,
+    const Tensor& input) {
+  gradInput.resize_as_(input);
```
Comment (skrah) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: The old code also zeros `gradInput`.

File: aten/src/ATen/native/native_functions.yaml [hunk 10]
```diff
@@ -3657,15 +3657,25 @@
...
 
 - func: adaptive_avg_pool3d(Tensor self, int[3] output_size) -> Tensor
   python_module: nn
+  dispatch:
+    CPU: adaptive_avg_pool3d_cpu
+    CUDA: adaptive_avg_pool3d_cuda
 
-- func: adaptive_avg_pool3d_backward(Tensor grad_output, Tensor self, *, Tensor(a!) grad_input) -> Tensor(a!)
-  python_module: nn
+- func: adaptive_avg_pool3d_backward(Tensor grad_output, Tensor self, *, Tensor(a!) gradInput) -> Tensor(a!)
```
Comment (skrah) [inline aten/src/ATen/native/native_functions.yaml]: It looks like the rest of the file uses `grad_input`.

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+
+  // the jit sometimes passes output_size.size() == 1
+  AT_CHECK(
+      output_size.size() == 1 || output_size.size() == 3,
+      "adaptive_avg_pool3d: internal error: output_size.size() must be 1 or 3");
+
+  int64_t osizeT = output_size[0];
+  int64_t osizeH = output_size[1];
+  int64_t osizeW = output_size[2];
+
```
Comment (skrah) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: If you allow `output.size.size() == 1`, you also have to set `osizeH` and `osizeW` conditionally.

But I think the `JIT` issue does not happen for `AdaptiveAveragePooling`, so just checking for `output_size.size() == 3` may be sufficient if the tests pass.

**Review by skrah — CHANGES_REQUESTED** [review]

**Comment by skrah** [comment]
@crcrpar I looks great to me in general, I've left some comments.

One completely general issue: Personally I'd find it easier to review if existing formatting is not changed during the porting step, for example `diff -uw existing.c new.cpp` is smaller.

But that is no change request, others PRs do it too.

**Comment by crcrpar** [comment]
@skrah Thanks for the review.

I've reflected your comments. I moved dispatch outside the loop  following `AdaptiveMaxPooling` implementation.

> One completely general issue: Personally I'd find it easier to review if existing formatting is not changed during the porting step, for example `diff -uw existing.c new.cpp` is smaller.

Sorry to bother you by this. I applied clang-format partially in this PR.

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 12]
```diff
@@ -215,22 +250,46 @@ static void adaptive_avg_pool3d_backward_out_frame(
...
   auto gradOutput = gradOutput_.contiguous();
 
+  /* sizes */
+  int64_t sizeD = input.size(-4);
+  int64_t isizeT = input.size(-3);
+  int64_t isizeH = input.size(-2);
+  int64_t isizeW = input.size(-1);
+  int64_t osizeT = gradOutput_.size(-3);
+  int64_t osizeH = gradOutput_.size(-2);
+  int64_t osizeW = gradOutput_.size(-1);
```
Comment (skrah) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: Again, extreme nitpick, but it still calls `gradOutput_.size(-3)` and not `gradOutput.size(-3)`. :)

File: aten/src/ATen/native/native_functions.yaml [hunk 13]
```diff
@@ -3657,15 +3657,25 @@
...
+    CUDA: adaptive_avg_pool3d_out_cuda
 
 - func: adaptive_avg_pool3d(Tensor self, int[3] output_size) -> Tensor
   python_module: nn
+  dispatch:
+    CPU: adaptive_avg_pool3d_cpu
+    CUDA: adaptive_avg_pool3d_cuda
 
 - func: adaptive_avg_pool3d_backward(Tensor grad_output, Tensor self, *, Tensor(a!) grad_input) -> Tensor(a!)
-  python_module: nn
```
Comment (skrah) [inline aten/src/ATen/native/native_functions.yaml]: I'm curious, why did you remove `python_module: nn` here but not above? Apparently it builds just fine, but most other locations in this file still have `python_module: nn`.

**Review by skrah — APPROVED** [review]

**Comment by skrah** [comment]
@crcrpar Thanks, I have added one additional question about `native_functions.yaml` and one small nitpick.

File: aten/src/ATen/native/native_functions.yaml (hunk 13 above)
Comment (crcrpar) [reply aten/src/ATen/native/native_functions.yaml]: Oh, this is my fault. I did not remove this deliberately and I get this back for consistency with others.

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp (hunk 12 above)
Comment (crcrpar) [reply aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: Thanks.

File: aten/src/ATen/native/native_functions.yaml [hunk 14]
```diff
@@ -3657,15 +3657,26 @@
...
+    CUDA: adaptive_avg_pool3d_cuda
 
 - func: adaptive_avg_pool3d_backward(Tensor grad_output, Tensor self, *, Tensor(a!) grad_input) -> Tensor(a!)
   python_module: nn
+  dispatch:
+    CPU: adaptive_avg_pool3d_backward_out_cpu
+    CUDA: adaptive_avg_pool3d_backward_out_cuda
 
 - func: adaptive_avg_pool3d_backward(Tensor grad_output, Tensor self) -> Tensor
-  python_module: nn
```
Comment (skrah) [inline aten/src/ATen/native/native_functions.yaml]: This one is still missing.

Comment (crcrpar) [reply aten/src/ATen/native/native_functions.yaml]: thanks... :bowing_man:

**Comment by skrah** [comment]
Thanks, looks great!   Let's ask @ezyang to take over.

**Comment by ezyang** [comment]
Thanks @skrah. As a methodological note, if you want to "accept" a review but pass it off to someone else for a complete review, instead of "accepting" it, just remove yourself from the reviewer list and add a comment LGTM. This prevents the PR as showing up as "accepted" in the UI, so that I know it does need review (in this case, I saved the email notification in my inbox for review) :)

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+namespace {
+
+inline int start_index(int a, int b, int c) {
+  return (int)std::floor((float)(a * c) / b);
+}
+
+inline int end_index(int a, int b, int c) {
+  return (int)std::ceil((float)((a + 1) * c) / b);
+}
+
```
Comment (ezyang) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: Did we lose this comment? Was it inaccurate?

```
/*
 * Description:
 *    This function adaptively average pools an input 5D tensor along dimensions
 *     2, 3 and 4.
 *
 *    gridDim.y blocks work together on a single 2D output plane specified by
 *    (blockIdx.x + offsetZ).
 */
```

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+    int64_t isizeW,
+    int64_t osizeT,
+    int64_t osizeH,
+    int64_t osizeW,
+    int64_t istrideD,
+    int64_t istrideT,
+    int64_t istrideH,
+    int64_t istrideW) {
+  int64_t d;
+#pragma omp parallel for private(d)
```
Comment (ezyang) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: This parallelization is new :) How much speed up did we get from it?

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu (hunk 11 above)
Comment (ezyang) [reply aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: Can we file an issue about this, and reference it to all of the cases where we added extra checks like this? I'm pretty sure this code is supposed to not be necessary, but since no one is root causing it, it looks like this code is going to replicate everywhere, can we at least put a comment on it.

**Comment by ezyang** [comment]
Everything looks in order, and review comments look addressable in follow up PRs.

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu (hunk 11 above)
Comment (skrah) [reply aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu]: Yes, it's a good idea to file an issue, I'll open one.  Probably for this file the workaround is not necessary, but I haven't verified.

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp (hunk 15 above)
Comment (crcrpar) [reply aten/src/ATen/native/AdaptiveAveragePooling3d.cpp]: Hmm, I guess you got a bit confused while reviewing because I found only the below comment in THNN code (C impl).
https://github.com/pytorch/pytorch/blob/cf55670bddf6bb5f6199df2c543c8fb4ca21b08e/aten/src/THNN/generic/VolumetricAdaptiveAveragePooling.c#L10

Also, the new `.cu` file has the comment of description: https://github.com/pytorch/pytorch/pull/19898/files#diff-b9ff310d15f69d0b85cfe7fd87d8f8b9R33

**Comment by crcrpar** [comment]
@ezyang Thanks for the look. Could you check comments (https://github.com/pytorch/pytorch/pull/19898#discussion_r281642657)?

**Comment by ezyang** [comment]
You're right!  My bad.

Excerpts from Masaki Kozuki's message of 2019-05-07 09:18:30 -0700:
> crcrpar commented on this pull request.
> 
> > +#include <ATen/ATen.h>
> +#include <ATen/NativeFunctions.h>
> +
> +namespace at {
> +namespace native {
> +
> +namespace {
> +
> +inline int start_index(int a, int b, int c) {
> +  return (int)std::floor((float)(a * c) / b);
> +}
> +
> +inline int end_index(int a, int b, int c) {
> +  return (int)std::ceil((float)((a + 1) * c) / b);
> +}
> +
> 
> Hmm, I guess you got a bit confused while reviewing because I found only the below comment in THNN code (C impl).
> https://github.com/pytorch/pytorch/blob/cf55670bddf6bb5f6199df2c543c8fb4ca21b08e/aten/src/THNN/generic/VolumetricAdaptiveAveragePooling.c#L10
> 
> Also, the new `.cu` file has the comment of description: https://github.com/pytorch/pytorch/pull/19898/files#diff-b9ff310d15f69d0b85cfe7fd87d8f8b9R33
>

**Comment by ezyang** [comment]
@crcrpar I saw you pushed new commits (which caused the in progress land to fail). Were there any changes?

**Comment by crcrpar** [comment]
@ezyang I found and fixed conflicts in `native_functions.yaml`.
